README.md
# [gpgpdump] - OpenPGP packet visualizer

[![Build Status](https://travis-ci.org/spiegel-im-spiegel/gpgpdump.svg?branch=master)](https://travis-ci.org/spiegel-im-spiegel/gpgpdump)
[![GitHub license](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/spiegel-im-spiegel/gpgpdump/master/LICENSE)
[![GitHub release](http://img.shields.io/github/release/spiegel-im-spiegel/gpgpdump.svg)](https://github.com/spiegel-im-spiegel/gpgpdump/releases/latest)

[gpgpdump] is a OpenPGP ([RFC 4880]) packet visualizer by [golang](https://golang.org/).

- Based on [pgpdump](https://github.com/kazu-yamamoto/pgpdump)
- Provide [golang](https://golang.org/) package and command-line Interface
- Output with [TOML](https://github.com/toml-lang/toml) (or [JSON](https://tools.ietf.org/html/rfc7159)) format
- Support [RFC 5581] and [RFC 6637]

## Install

```
$ go get -v github.com/spiegel-im-spiegel/gpgpdump
```

Installing by [dep].

```
$ dep ensure -add github.com/spiegel-im-spiegel/godump
```

### Usage

```go
const openpgpStr = `
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iF4EARMIAAYFAlTDCN8ACgkQMfv9qV+7+hg2HwEA6h2iFFuCBv3VrsSf2BREQaT1
T1ZprZqwRPOjiLJg9AwA/ArTwCPz7c2vmxlv7sRlRLUI6CdsOqhuO1KfYXrq7idI
=ZOTN
-----END PGP SIGNATURE-----
`

info, err := gpgpdump.Parse(strings.NewReader(openpgpStr), options.New())
if err != nil {
	return
}
fmt.Println(info.Packets[0].Value)
// Output:
// Signature Packet (tag 2)

```

## Command-Line Interface

### Binaries

See [latest release](https://github.com/spiegel-im-spiegel/gpgpdump/releases/latest).

### Usage

```
$ gpgpdump -h
Usage:
  gpgpdump [flags] [PGPfile]

Flags:
  -a, --armor     accepts ASCII input only
  -h, --help      help for gpgpdump
  -i, --int       dumps multi-precision integers
  -j, --json      output with JSON format
  -l, --literal   dumps literal packets (tag 11)
  -m, --marker    dumps marker packets (tag 10)
  -p, --private   dumps private packets (tag 60-63)
  -u, --utc       output with UTC time
  -v, --version   output version of gpgpdump

$ cat sig
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iF4EARMIAAYFAlTDCN8ACgkQMfv9qV+7+hg2HwEA6h2iFFuCBv3VrsSf2BREQaT1
T1ZprZqwRPOjiLJg9AwA/ArTwCPz7c2vmxlv7sRlRLUI6CdsOqhuO1KfYXrq7idI
=ZOTN
-----END PGP SIGNATURE-----

$ cat sig | gpgpdump -u
[[Packet]]
  name = "Packet"
  value = "Signature Packet (tag 2)"
  note = "94 bytes"

  [[Packet.Item]]
    name = "Version"
    value = "4"
    note = "new"

  [[Packet.Item]]
    name = "Signiture Type"
    value = "Signature of a canonical text document (0x01)"

  [[Packet.Item]]
    name = "Public-key Algorithm"
    value = "ECDSA public key algorithm (pub 19)"

  [[Packet.Item]]
    name = "Hash Algorithm"
    value = "SHA256 (hash 8)"

  [[Packet.Item]]
    name = "Hashed Subpacket"
    note = "6 bytes"

    [[Packet.Item.Item]]
      name = "Signature Creation Time (sub 2)"
      value = "2015-01-24T02:52:15Z"

  [[Packet.Item]]
    name = "Unhashed Subpacket"
    note = "10 bytes"

    [[Packet.Item.Item]]
      name = "Issuer (sub 16)"
      value = "0x31fbfda95fbbfa18"

  [[Packet.Item]]
    name = "Hash left 2 bytes"
    dump = "36 1f"

  [[Packet.Item]]
    name = "Multi-precision integer"
    note = "ECDSA r (256 bits)"

  [[Packet.Item]]
    name = "Multi-precision integer"
    note = "ECDSA s (252 bits)"

$ cat sig | gpgpdump -j -u
{
  "Packet": [
    {
      "name": "Packet",
      "value": "Signature Packet (tag 2)",
      "note": "94 bytes",
      "Item": [
        {
          "name": "Version",
          "value": "4",
          "note": "new"
        },
        {
          "name": "Signiture Type",
          "value": "Signature of a canonical text document (0x01)"
        },
        {
          "name": "Public-key Algorithm",
          "value": "ECDSA public key algorithm (pub 19)"
        },
        {
          "name": "Hash Algorithm",
          "value": "SHA256 (hash 8)"
        },
        {
          "name": "Hashed Subpacket",
          "note": "6 bytes",
          "Item": [
            {
              "name": "Signature Creation Time (sub 2)",
              "value": "2015-01-24T02:52:15Z"
            }
          ]
        },
        {
          "name": "Unhashed Subpacket",
          "note": "10 bytes",
          "Item": [
            {
              "name": "Issuer (sub 16)",
              "value": "0x31fbfda95fbbfa18"
            }
          ]
        },
        {
          "name": "Hash left 2 bytes",
          "dump": "36 1f"
        },
        {
          "name": "Multi-precision integer",
          "note": "ECDSA r (256 bits)"
        },
        {
          "name": "Multi-precision integer",
          "note": "ECDSA s (252 bits)"
        }
      ]
    }
  ]
}
```

## Dependencies

```
dep status -dot | dot -Tpng -o dependency.png
```

[![Dependencies](dependency.png)](dependency.png)

[gpgpdump]: https://github.com/spiegel-im-spiegel/gpgpdump "spiegel-im-spiegel/gpgpdump: gpgpdump - OpenPGP packet visualizer"
[RFC 4880]: https://tools.ietf.org/html/rfc4880
[RFC 5581]: http://tools.ietf.org/html/rfc5581
[RFC 6637]: http://tools.ietf.org/html/rfc6637
[dep]: https://github.com/golang/dep "golang/dep: Go dependency management tool"


CHANGES.md
# Change Log

## 0.2.0 2017-11-07

- Rewrite all codes for v0.2

## 0.1.2 2016-02-13

- Many refactoring & debug.
- Distribute binaries. (see [latest release](https://github.com/spiegel-im-spiegel/gpgpdump/releases/latest))
	- add scripts folder and sh script file.

## 0.1.1 2016-02-12

- Many refactoring & debug.

## 0.1.0 2016-02-07

- The first release.


options.go
package gpgpdump

import "fmt"

const (
	//ArmorOpt is name of armor option
	ArmorOpt = "armor"
	//DebugOpt is name of debug option
	DebugOpt = "debug"
	//GDumpOpt is name of gdump option
	GDumpOpt = "gdump"
	//IntegerOpt is name of int option
	IntegerOpt = "int"
	//JSONOpt is name of json option
	JSONOpt = "json"
	//LiteralOpt is name of literal option
	LiteralOpt = "literal"
	//MarkerOpt is name of marker option
	MarkerOpt = "marker"
	//PrivateOpt is name of private option
	PrivateOpt = "private"
	//UTCOpt is name of utc option
	UTCOpt = "utc"
)

// Options for gpgpdump
type Options struct {
	armorFlag   bool //accepts ASCII input only
	debugFlag   bool //for debug
	gdumpFlag   bool //selects alternate (GnuPG type) dump format (not used)
	intFlag     bool //dumps multi-precision integers
	jsonFlag    bool //output with JSON format
	literalFlag bool //dumps literal packets (tag 11)
	markerFlag  bool //dumps marker packets (tag 10)
	privateFlag bool //dumps private packets (tag 60-63)
	utcFlag     bool //output UTC time
}

//OptFunc is self-referential function for functional options pattern
type OptFunc func(*Options)

// NewOptions returns a new UI instance
func NewOptions(opts ...OptFunc) *Options {
	o := &Options{}
	o.options(opts...)
	return o
}
func (o *Options) options(opts ...OptFunc) {
	for _, opt := range opts {
		opt(o)
	}
}

//Set returns closure as type OptFunc
func Set(name string, f bool) OptFunc {
	return func(o *Options) {
		o.Set(name, f)
	}
}

//Set sets option to Options.
func (o *Options) Set(name string, f bool) {
	switch name {
	case ArmorOpt:
		o.armorFlag = f
	case DebugOpt:
		o.debugFlag = f
	case GDumpOpt:
		o.gdumpFlag = f
	case IntegerOpt:
		o.intFlag = f
	case JSONOpt:
		o.jsonFlag = f
	case LiteralOpt:
		o.literalFlag = f
	case MarkerOpt:
		o.markerFlag = f
	case PrivateOpt:
		o.privateFlag = f
	case UTCOpt:
		o.utcFlag = f
	}
}

//Armor return flag value of armorFlag
func (o *Options) Armor() bool { return o.armorFlag }

//Debug return flag value of debugFlag
func (o *Options) Debug() bool { return o.debugFlag }

//GDump return flag value of gdumpFlag
func (o *Options) GDump() bool { return o.gdumpFlag }

//Integer return flag value of intFlag
func (o *Options) Integer() bool { return o.intFlag }

//JSON return flag value of jsonFlag
func (o *Options) JSON() bool { return o.jsonFlag }

//Literal return flag value of literalFlag
func (o *Options) Literal() bool { return o.literalFlag }

//Marker return flag value of markerFlag
func (o *Options) Marker() bool { return o.markerFlag }

//Private return flag value of privateFlag
func (o *Options) Private() bool { return o.privateFlag }

//UTC return flag value of utcFlag
func (o *Options) UTC() bool { return o.utcFlag }

//Stringer
func (o *Options) String() string {
	return fmt.Sprint(
		ArmorOpt, ":", o.Armor(), ",",
		DebugOpt, ":", o.Debug(), ",",
		GDumpOpt, ":", o.GDump(), ",",
		IntegerOpt, ":", o.Integer(), ",",
		JSONOpt, ":", o.JSON(), ",",
		LiteralOpt, ":", o.Literal(), ",",
		MarkerOpt, ":", o.Marker(), ",",
		PrivateOpt, ":", o.Private(), ",",
		UTCOpt, ":", o.UTC())
}

/* Copyright 2017 Spiegel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
